File: week13/20/pr5_maximum_difference_between_node_and_ancestor.c

//  * Definition for a binary tree node.
  struct TreeNode {
      int val;
      struct TreeNode *left;
      struct TreeNode *right;
  };

int max(int a, int b){
    return a > b ? a : b;
}
int min(int a, int b){
    return a > b? b : a;
}

int helper(struct TreeNode* node, int maxval, int minval){
    if(!node) return maxval - minval;
    maxval = max(maxval,node->val);
    minval = min(minval,node->val);
    int left = helper(node->left,maxval,minval);
    int right = helper(node->right,maxval,minval);
    return max(left,right);
}

int maxAncestorDiff(struct TreeNode* root) {
    return helper(root,root->val,root->val);
}

File: week13/20/pr5_maximum_difference_between_node_and_ancestor.py
# test = [1,None,3]

# print(max(1,0))
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def maxAncestorDiff(self, root: TreeNode) -> int:
        def dfs(node,curr_max,curr_min):
            if not node : return 0

            curr_max = max(curr_max,node.val)
            curr_min = min(curr_min,node.val)

            if not node.left and not node.right :
                return curr_max - curr_min

            left = dfs(node.left,curr_max,curr_min)
            right = dfs(node.right,curr_max,curr_min)

            return max(left,right)

        return dfs(root,root.val,root.val)